docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,519 @@
+      uint32_t log_level,
+      const char* message)
+    {
+        oe_result_t _result = OE_FAILURE;
+
+        /* CRC32 of oe_log_ocall */
+        const uint32_t _function_hash = 3195194105;
+
+        /* Used to cache the ocall id. */
+        static int _ocall_id = OE_OCALL_ID_NULL;
+        
+        ...
+        
+        /* Check if the ocall id is cached. */
+        if (_ocall_id == OE_OCALL_ID_NULL)
+          _ocall_id = oe_host_get_ocall_id_by_hash(_function_hash);
+
+        /* Call host function. */
+        if ((_result = oe_call_host_function(
+             (uint64_t) _ocall_id,
+             (uint64_t) _function_hash,
+             _input_buffer,
+             _input_buffer_size,
+             _output_buffer,
+             _output_buffer_size,
+             &_output_bytes_written)) != OE_OK)
+            goto done;
+
+        ...
+     }
+  ```
+
+- Caching an ECALL id.
+  
+  Because a single host may interact with multiple enclaves, we cannot directly cache the ECALL id inside an ECALL wrapper;
+  i.e., the host may use the same ECALL wrapper to dispatch the request to different enclaves. As a result, we need to
+  cache the id per enclave. More specifically, the host maintains an id caching table of ECALLs per `enclave` struct. Note that
+  this table includes all the ECALLs from multiple enclaves such that every `enclave` structs have the same table.
+  
+  Take a similar example of two enclaves: `enclave_foo` and `enclave_bar`. Both of the enclaves implement
+  the ECALL `sample_ecall`, but with different ids (e.g., `1` in `enclave_foo` and `2` in `enclave_bar`).
+  To cache the id per enclave, the `enclave` structs of both enclaves, `enclave_foo` and `_enclave_bar`,
+  on the host maintains the same table `ecall_id_table`. As a result, the index of `sample_ecall` in
+  `enclave_foo->ecall_id_table` and `enclave_bar->ecall_id_table` is the same, which allows for caching
+  the ECALL id on the per enclave basis.
+  
+  With this, the ECALL wrapper looks up the id to the per-enclave caching table and then obtains
+  the ECALL id from the enclave (via a specialized ECALLs) for the first invocation.
+  In the subsequent invocations, the ECALL wrapper uses the cached caching table id and per-enclave
+  ECALL id to make the ECALL. See the following code snippet for an example.
+  
+  ```
+  oe_result_t oe_verify_report_ecall(
+    oe_enclave_t* enclave,
+    oe_result_t* _retval,
+    const void* report,
+    size_t report_size)
+  {
+      oe_result_t _result = OE_FAILURE;
+
+      /* CRC32 of oe_verify_report_ecall */
+      const uint32_t _function_hash = 3802944385;
+
+      /* Used to cache the index to the id caching table. */
+      static int _caching_table_id = OE_HOST_ECALL_ID_NULL;
+
+      int _ecall_id = -1;
+      
+      ...
+      
+      /* Check if the index is cached. */
+      if (_index == OE_HOST_ECALL_ID_NULL)
+      {
+        _index = oe_host_get_ecall_id_by_hash(
+            enclave,
```

Comment:
Since
> As a result, the index of `sample_ecall` in
  `enclave_foo->ecall_id_table` and `enclave_bar->ecall_id_table` is the same, which allows for caching
  the ECALL id on the per enclave basis.

do we need the enclave as a parameter here?

Reply:
Yes. This function call is essentially an ECALL. We need to specify the `enclave` here.